Add a compatibility layer with database/sql in the standard library

kkoreilly opened the pull request:
I understand the reasons stated in the readme on why a `database/sql` driver was not made. However, although directly using a gorqlite-specific API provides a better experience, an additional `database/sql` driver would significantly broaden the reach of gorqlite by allowing it to be used with other libraries like GORM. In fact, the main reason I wrote this PR is so that I can write a GORM driver for gorqlite.

This PR introduces a simple wrapper library, `stdlib`, which implements `database/sql/driver` interfaces. The implementations are trivial wrappers of already existing functions in the main `gorqlite` package, and no functionality in the main package is lost. This structure is based on [jackc/pgx](https://github.com/jackc/pgx), which is a popular Go PostgreSQL driver.

There are only two changes to the main package in this PR:

1. The addition of `QueryResult.Slice`, which is necessary because `database/sql/driver.Rows.Next` needs to copy the values into a destination slice of `driver.Value` objects. It is based on `QueryResult.Map`, and has effectively the same API and code, just with a slice instead of a map.
2. A simple if statement in `Connection.rqliteApiCall` which ensures that a nil request body reader is passed for a nil request body. This is necessary because browsers are very strict about having empty bodies on `GET` requests, so an empty but non-nil buffer results in all gorqlite HTTP `GET` requests failing in browsers.

otoolep commented on stdlib/sql.go:
```diff
@@ -0,0 +1,171 @@
+	return nil
+}
+
+func (c *Conn) Begin() (driver.Tx, error) {
```
Should the return something like "not supported" since rqlite doesn't support traditional transactions?

kkoreilly replied:
Many operations in database/sql call this function, so it must return a usable value and not an error. It does not actually do anything, since rqlite does not support this kind of transaction, but it must be there as a placeholder that satisfies interfaces. In other words, as soon as Exec or Query is called, changes are instantly made, so all of the transaction functions are just no-ops, but they must exist.

This is the principle behind this PR: database/sql is an unideal wrapper for rqlite (which leads to many no-op functions), but it is still a perfectly functional wrapper. Therefore, given that it is the widely accepted standard in the Go community that everything else integrates with, I think that it makes sense to provide a compatibility layer for it.

otoolep commented on query.go:
```diff
@@ -388,6 +388,39 @@ func (qr *QueryResult) Map() (map[string]interface{}, error) {
+// The slice is a shallow copy of the internal representation of the row data.
+//
+// Note that only json values are supported, so you will need to type the interface{} accordingly.
+func (qr *QueryResult) Slice() ([]interface{}, error) {
```
Can you add a unit test for this function?

otoolep commented on README.md:
```diff
@@ -212,6 +212,25 @@ If you use access control, any user connecting will need the "status" permission
 
+## Using with database/sql
+
+It is recommended that you use the main gorqlite-specific API when possible. However, if you need to use gorqlite with database/sql, you can just import `github.com/rqlite/gorqlite/stdlib`. For example:
```
Do we need to highlight the limitations?

otoolep commented on stdlib/sql.go:
```diff
@@ -0,0 +1,171 @@
+// Package stdlib provides a compatability layer from gorqlite to database/sql.
+package stdlib
```
Testing is important to me -- don't we need unit tests for these functions?

otoolep reviewed (commented):
Thanks very much for this -- this is very interesting.

I believe it needs more test coverage however.

kkoreilly commented:
Of course. I agree that testing is very important, and I will add tests for all of the new functions and improve the documentation in the README.

kkoreilly commented:
I have addressed everything you mentioned. I do not know why the tests are failing; if you look at the history, they were passing before and then stopped passing when I changed the README, which shouldn't have any effect. Looking at the history of the main repository, it seems like the tests have randomly failed a couple of times recently, so there may be some problem making them not deterministic.

otoolep commented:
I can dig into the test failure, but it's failing in code you made a change to, right? By its nature of such a failure, it just happened to come out when the README change ran.

```
=== NAME  TestQueryOne
    query_test.go:41: failed during dropping table: there were 1 statement errors
    query_test.go:44: caught error: there were 1 statement errors
--- FAIL: TestQueryOne (0.01s)
    --- PASS: TestQueryOne/Normal_case (0.00s)
    --- PASS: TestQueryOne/Invalid_Query (0.00s)
    --- FAIL: TestQueryOne/Map_before_next (0.00s)
panic: runtime error: index out of range [0] with length 0 [recovered]
        panic: runtime error: index out of range [0] with length 0

goroutine 24 [running]:
testing.tRunner.func1.2({0x712840, 0xc000296f30})
        /usr/local/go/src/testing/testing.go:1526 +0x24e
testing.tRunner.func1()
        /usr/local/go/src/testing/testing.go:1529 +0x39f
panic({0x712840, 0xc000296f30})
        /usr/local/go/src/runtime/panic.go:884 +0x213
github.com/rqlite/gorqlite.(*QueryResult).Map(0xc000155e80)
        /home/circleci/rqlite/src/github.com/rqlite/gorqlite/query.go:370 +0x3e5
github.com/rqlite/gorqlite_test.TestQueryOne.func4(0xc0001036c0)
        /home/circleci/rqlite/src/github.com/rqlite/gorqlite/query_test.go:199 +0x1f2
testing.tRunner(0xc0001036c0, 0x7617c0)
        /usr/local/go/src/testing/testing.go:1576 +0x10b
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:1629 +0x3ea
FAIL    github.com/rqlite/gorqlite      0.023s
=== RUN   TestDisableClusterDiscovery
--- PASS: TestDisableClusterDiscovery (0.00s)
=== RUN   TestEnableClusterDiscovery
--- PASS: TestEnableClusterDiscovery (0.00s)
PASS
ok      github.com/rqlite/gorqlite/integration  0.006s
=== RUN   TestTable
    sql_test.go:68: trying Exec CREATE TABLE
=== RUN   TestTable/Exec_INSERT
=== RUN   TestTable/Query_SELECT
=== RUN   TestTable/Invalid_Query
--- PASS: TestTable (0.02s)
    --- PASS: TestTable/Exec_INSERT (0.01s)
    --- PASS: TestTable/Query_SELECT (0.00s)
    --- PASS: TestTable/Invalid_Query (0.00s)
PASS
```
Is it possible something you changed has introduced a race into the code?

otoolep commented:
>Looking at the history of the main repository, it seems like the tests have randomly failed a couple of times recently,

Can you show me where you're seeing that?

kkoreilly commented:
I fixed the tests. The new stdlib tests were using the same table as the main tests, so they were creating, deleting, modifying, and reading the table at the same time. I changed the default table name for the stdlib tests to `gorqlite_test_stdlib` and added the `GORQLITE_TEST_TABLE_STDLIB` env variable for changing it.

otoolep commented:
Looks good -- thanks for your contribution.

kkoreilly commented:
Yes, I meant to link to the GORM driver at some point. It is currently here: https://github.com/goki/rqlite, but I might move it somewhere else soon. It is fully functional.
